base/src/main/java/proguard/optimize/kotlin/KotlinLambdaMerger.java

```diff
@@ -0,0 +1,440 @@
+     * @param programClass the class of which the arity must be determined
+     * @return the value following the base name "kotlin/jvm/functions/Function". This can be a number going from
+     *         0 to 22 or the character "N". For example, for a class implementing the interface
+     *         "kotlin/jvm/functions/Function5" the value "5" is returned.
+     * @throws IllegalArgumentException if the given programClass does not implement a Kotlin function interface.
+     */
+    public static String getArityFromInterface(ProgramClass programClass) throws IllegalArgumentException
+    {
+        for (int interfaceIndex = 0; interfaceIndex < programClass.u2interfacesCount; interfaceIndex++)
+        {
+            String interfaceName = programClass.getInterfaceName(interfaceIndex);
+            if (interfaceName.startsWith(NAME_KOTLIN_FUNCTION) && interfaceName.length() > NAME_KOTLIN_FUNCTION.length())
+            {
+                return interfaceName.substring(NAME_KOTLIN_FUNCTION.length());
+            }
+        }
+        throw new IllegalArgumentException("Class " + ClassUtil.externalClassName(programClass.getName()) +
+                                           " does not implement a Kotlin function interface.");
+    }
+
+    /**
+     * Checks whether the given lambda class should still be merged.
+     * Returns true if the lambda class has not yet been merged and is allowed to be merged.
+     * @param lambdaClass the lambda class for which should be checked whether it should be merged
+     */
+    public static boolean shouldMerge(ProgramClass lambdaClass)
+    {
+        ProgramClassOptimizationInfo optimizationInfo =
+                ProgramClassOptimizationInfo.getProgramClassOptimizationInfo(lambdaClass);
+        return (lambdaClass.getProcessingFlags() & (ProcessingFlags.DONT_OPTIMIZE | ProcessingFlags.DONT_SHRINK)) == 0
+                && lambdaClass.extendsOrImplements(NAME_KOTLIN_LAMBDA)
+                // if optimisation info is null, then the lambda was not enqueued to be merged
+                && optimizationInfo != null
+                && optimizationInfo.getLambdaGroup() == null
+                && optimizationInfo.mayBeMerged();
+    }
+
+    /**
+     * Checks whether the given lambda class can be merged and throws an exception if it cannot be merged.
+     * @param lambdaClass the lambda class for which it should be checked whether it can be merged
+     * @param mergeLambdaClassesWithUnexpectedMethods a configuration setting that allows merging lambda classes that
+ *                                                    contain unexpected instance methods
+     * @param programClassPool the program class pool in which the lambda class is defined and used
+     * @throws IllegalArgumentException if the given lambda class cannot be merged
+     */
+    public static void ensureCanMerge(ProgramClass lambdaClass,
+                                      boolean mergeLambdaClassesWithUnexpectedMethods,
```

Comment:
What is an unexpected method? When would someone chose to use this option? Can it cause any problems?